src/Factory/ServerRequestFactory.php

```diff
@@ -4,35 +4,184 @@
 
 namespace Nyholm\Psr7\Factory;
 
+use InvalidArgumentException;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Message\UploadedFileInterface;
 use Interop\Http\Factory\ServerRequestFactoryInterface;
 use Nyholm\Psr7\ServerRequest;
+use Nyholm\Psr7\Factory\UriFactory;
+use Nyholm\Psr7\UploadedFile;
 
 /**
  * @author Tobias Nyholm <tobias.nyholm@gmail.com>
  * @author Martijn van der Ven <martijn@vanderven.se>
  */
 class ServerRequestFactory implements ServerRequestFactoryInterface
 {
-    public function createServerRequest($method, $uri)
+    public function createServerRequest($method, $uri): ServerRequestInterface
     {
         return new ServerRequest($method, $uri);
     }
 
-    public function createServerRequestFromArray(array $server)
+    public function createServerRequestFromArray(array $server): ServerRequestInterface
     {
-        if (!isset($server['REQUEST_METHOD'])) {
-            throw new \InvalidArgumentException('Cannot determine HTTP method');
+        return new ServerRequest(
+            $this->getMethodFromEnvironment($server),
+            $this->getUriFromEnvironmentWithHTTP($server)
+        );
+    }
+
+    /**
+     * Create a new server request from a set of arrays.
+     *
+     * @param array $server  Typically $_SERVER or similar structure.
+     * @param array $headers Typically the output of getallheaders() or similar structure.
+     * @param array $cookie  Typically $_COOKIE or similar structure.
+     * @param array $get     Typically $_GET or similar structure.
+     * @param array $post    Typically $_POST or similar structure.
+     * @param array $files   Typically $_FILES or similar structure.
+     *
+     * @throws InvalidArgumentException If no valid method or URI can be determined.
+     *
+     * @return ServerRequestInterface
+     */
+    public function createServerRequestFromArrays(
```

Comment:
> Hm, how do you feel about this name?

I have no strong feelings about it. I was merely following the factory’s `createServerRequestFromArray`. PSR-17 has a good rationale for including “ServerRequest” in the method names, though that doesn’t apply to us.

I can change it to anything you’d like, really.